### PR TITLE
introduce TensorBase::mutable_data_ptr()

### PR DESCRIPTION
**Pull request by dagitses**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #98162
* __->__ #97859
* #97874
* #97804
* #97776
* #97744
* #97743
* #97694
* #97690
* #98152

See D44409928 for motivation.

Note that we keep the const-ness of the existing data_ptr() member so
that we don't have to change all references atomically. We just change
the ones here that we have higher confidence with.

Differential Revision: [D44492539](https://our.internmc.facebook.com/intern/diff/D44492539/)

**NOTE FOR REVIEWERS**: This PR has internal Meta-specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D44492539/)!

cc @ezyang @bhosmer @smessmer @ljk53 @bdhirsh @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @EikanWang

### CODE REVIEW

File: aten/src/ATen/cuda/CUDASparseDescriptors.cpp [hunk 1]
```diff
@@ -81,7 +81,7 @@ CuSparseDnMatDescriptor::CuSparseDnMatDescriptor(const Tensor& input, int64_t ba
 #endif
 
   auto batch_stride = ndim > 2 && batch_offset >= 0 ? input_strides[ndim - 3] : 0;
-  void* values_ptr = static_cast<char*>(input.data_ptr()) +
+  void* values_ptr = static_cast<char*>(input.mutable_data_ptr()) +
```
Comment (ezyang) [inline aten/src/ATen/cuda/CUDASparseDescriptors.cpp:84]: This looks wrong. Based on the description here I doubt this pointer is actually getting mutated. Generally, you can't trust the underlying libraries to have correct const or not types

File: aten/src/ATen/cuda/CUDASparseDescriptors.cpp [hunk 2]
```diff
@@ -121,7 +121,7 @@ CuSparseDnVecDescriptor::CuSparseDnVecDescriptor(const Tensor& input) {
 
   cusparseDnVecDescr_t raw_descriptor;
   TORCH_CUDASPARSE_CHECK(cusparseCreateDnVec(
-      &raw_descriptor, input.numel(), input.data_ptr(), value_type));
+      &raw_descriptor, input.numel(), input.mutable_data_ptr(), value_type));
```
Comment (ezyang) [inline aten/src/ATen/cuda/CUDASparseDescriptors.cpp:124]: Ditto

File: aten/src/ATen/cuda/CUDASparseDescriptors.cpp [hunk 3]
```diff
@@ -163,13 +163,13 @@ CuSparseSpMatCsrDescriptor::CuSparseSpMatCsrDescriptor(const Tensor& input, int6
...
-      static_cast<char*>(crow_indices.data_ptr()) +
+      static_cast<char*>(crow_indices.mutable_data_ptr()) +
           batch_offset * crow_indices_batch_stride * crow_indices.itemsize(),
       // column indices of the sparse matrix, size = nnz
-      static_cast<char*>(col_indices.data_ptr()) +
+      static_cast<char*>(col_indices.mutable_data_ptr()) +
           batch_offset * col_indices_batch_stride * col_indices.itemsize(),
       // values of the sparse matrix, size = nnz
-      static_cast<char*>(values.data_ptr()) +
+      static_cast<char*>(values.mutable_data_ptr()) +
```
Comment (ezyang) [inline aten/src/ATen/cuda/CUDASparseDescriptors.cpp:172]: Ditto x3

File: aten/src/ATen/cuda/CUDASparseDescriptors.h [hunk 4]
```diff
@@ -195,9 +195,9 @@ class TORCH_CUDA_CPP_API CuSparseSpMatCsrDescriptor
     TORCH_INTERNAL_ASSERT_DEBUG_ONLY(values.is_contiguous());
     TORCH_CUDASPARSE_CHECK(cusparseCsrSetPointers(
         this->descriptor(),
-        crow_indices.data_ptr(),
-        col_indices.data_ptr(),
-        values.data_ptr()));
+        crow_indices.mutable_data_ptr(),
+        col_indices.mutable_data_ptr(),
+        values.mutable_data_ptr()));
```
Comment (ezyang) [inline aten/src/ATen/cuda/CUDASparseDescriptors.h:200]: Ditto

File: aten/src/ATen/cudnn/Descriptors.h [hunk 5]
```diff
@@ -227,14 +227,14 @@ struct TORCH_CUDA_CPP_API DropoutDescriptor
     AT_ASSERT(options.device().type() == kCUDA);
     AT_ASSERT(options.dtype() == kByte);
     state = at::empty({static_cast<int64_t>(state_size)}, options);
-    AT_CUDNN_CHECK(cudnnSetDropoutDescriptor(mut_desc(), handle, dropout, state.data_ptr(), state_size, seed));
+    AT_CUDNN_CHECK(cudnnSetDropoutDescriptor(mut_desc(), handle, dropout, state.mutable_data_ptr(), state_size, seed));
```
Comment (ezyang) [inline aten/src/ATen/cudnn/Descriptors.h:230]: Ditto

File: aten/src/ATen/cudnn/Descriptors.h [hunk 6]
```diff
@@ -227,14 +227,14 @@ struct TORCH_CUDA_CPP_API DropoutDescriptor
...
   }
 
   // Restore a dropout descriptor given a dropout probability and existing RNG state.
   void set(cudnnHandle_t handle, float dropout, at::Tensor state_) {
     TORCH_INTERNAL_ASSERT(dropout > 0, "dropout must be nonzero; otherwise call set_no_dropout");
     state = state_;
-    void *state_ptr = state.data_ptr();
+    void *state_ptr = state.mutable_data_ptr();
```
Comment (ezyang) [inline aten/src/ATen/cudnn/Descriptors.h:237]: Not sure about this one as dropout state may get subsequently mutated through the descriptor. You will need to read docs

File: aten/src/ATen/native/Embedding.cpp [hunk 7]
```diff
@@ -129,8 +129,8 @@ Tensor embedding_dense_backward_cpu(
     .declare_static_shape(grad.sizes(), /*squash_dims=*/0)
     .build();
 
-  const auto gW_data = reinterpret_cast<char*>(grad_weight.data_ptr());
-  const auto gO_data = reinterpret_cast<char*>(grad.data_ptr());
```
Comment (ezyang) [inline aten/src/ATen/native/Embedding.cpp:133]: gO should be read only (gW should be write)

**Comment by ezyang** [comment]
I'm pausing review, please do a reaudit of the rest of your changes. We may also want to discuss what the API for "I promise this is non mutating but the downstream API is not const correct" (most straightforward is to use const_cast, albeit a bit wordy)

**Comment by dagitses** [comment]
> 

I propose the following:
* in the short-term, we have `tensor.must_audit_mutable_data_ptr()`. This can just reflect the status quo of what we have an gives us a convenient string to search for.
* regarding const_cast, it might be nice to have a debug-only API that helps verify the mutable borrow is not mutated, something like:

```
{
  auto data_ptr = tensor.borrow_const_data_ptr();
  someCudaApi(data_ptr.subtle_as_non_const_for_naughty_api());
}
```

Here the result of `borrow_const_data_ptr()` is an RAII object that hashes the data upon construction in debug mode and hashes on destruction in debug mode and asserts they are identical.

Even if we deem that too excessive, I would vote in favor of having a member function that is explicit about what exactly we are doing, e.g. `unsafe_get_non_const_data_ptr_for_non_const_correct_api()`. Otherwise, I would feel compelled to add a wordy comment any place I used `const_cast` to justify it.

Personally, I like the idea of the debug-only hashing. Trust but verify.

**Comment by ezyang** [comment]
A dispatcher level check could look something like this: upon entering a function, we setup some TLS mapping input tensors to whether or not they showed up in mutable or non-mutable argument positions according to schema. We now error if you access a data pointer on a tensor that is not explicitly mentioned, or access mutable data pointer on a non-mutable argument.

This check is unlikely to play well with `__torch_dispatch__` though, so it will take some thought on how to design it correctly. See also @albanD constantly having to fend off people who want to delete the refcount asserts from autograd.

**Comment by dagitses** [comment]
> A dispatcher level check could look something like this: upon entering a function, we setup some TLS mapping input tensors to whether or not they showed up in mutable or non-mutable argument positions according to schema. We now error if you access a data pointer on a tensor that is not explicitly mentioned, or access mutable data pointer on a non-mutable argument.
> 
> This check is unlikely to play well with `__torch_dispatch__` though, so it will take some thought on how to design it correctly. See also @albanD constantly having to fend off people who want to delete the refcount asserts from autograd.

That seems reasonable to me. OK, that's not going to happen in the very short-term. So what do you think about just having a naming convention to address the two cases we're concerned about here:
* `tensor.get_data_as_non_const_for_external_api()`: for where we'd use `const_cast`
* `tensor.must_audit_get_mutable_data()`: for where we want to move off of the const returning accessor but need to more closely scrutinize in the medium term.

For the latter, I'm primarily concerned with being able to communicate areas of uncertainty to local experts. For example, it would be nice to put those changes into the flash attention implementation and defer to Driss to resolve them. I don't want to have to make this call authoritatively for every single function.

**Comment by ezyang** [comment]
ok

**Comment by dagitses** [comment]
> ok

OK, I think I have an even better idea for the first problem. How about we wrap the "unsafe" APIs with const correct wrappers? Then we do the `const_cast` inside the wrapper. This makes the most sense to me because we should be auditing APIs for const correctness, not callers.

**Comment by ezyang** [comment]
I'm cool with that too.
